leveldb close open iterators error on shutdown

shutting down a local node following fusaka devnet 3. Seems happening regularly. 

```
2025-09-15 21:39:26.693 INFO  - Shutting down discovery server listening on /0.0.0.0:9000
2025-09-15 21:39:28.757 ERROR - Failed to stop services
java.util.concurrent.CompletionException: java.lang.NegativeArraySizeException: -39
at java.base/java.util.concurrent.CompletableFuture.wrapInCompletionException(CompletableFuture.java:325) ~[?:?]
at java.base/java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:414) ~[?:?]
at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1235) ~[?:?]
at java.base/java.util.concurrent.CompletableFuture.thenCompose(CompletableFuture.java:2393) ~[?:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.thenCompose(SafeFuture.java:651) ~[classes/:?]
at tech.pegasys.teku.services.chainstorage.StorageService.doStop(StorageService.java:317) ~[classes/:?]
at tech.pegasys.teku.service.serviceutils.Service.stop(Service.java:44) ~[classes/:?]
at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:215) ~[?:?]
at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1716) ~[?:?]
at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570) ~[?:?]
at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560) ~[?:?]
at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:636) ~[?:?]
at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:291) ~[?:?]
at java.base/java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:656) ~[?:?]
at tech.pegasys.teku.services.ServiceController.doStop(ServiceController.java:39) ~[classes/:?]
at tech.pegasys.teku.service.serviceutils.Service.stop(Service.java:44) ~[classes/:?]
at tech.pegasys.teku.AbstractNode.stop(AbstractNode.java:251) ~[classes/:?]
at tech.pegasys.teku.Teku.lambda$main$0(Teku.java:49) ~[classes/:?]
at java.base/java.lang.Thread.run(Thread.java:1447) [?:?]
Caused by: java.lang.NegativeArraySizeException: -39
at java.base/java.util.HashSet.toArray(HashSet.java:377) ~[?:?]
at java.base/java.util.ArrayList.<init>(ArrayList.java:182) ~[?:?]
at tech.pegasys.teku.storage.server.leveldb.LevelDbInstance.close(LevelDbInstance.java:336) ~[classes/:?]
at tech.pegasys.teku.storage.server.kvstore.dataaccess.CombinedKvStoreDao.close(CombinedKvStoreDao.java:264) ~[classes/:?]
at tech.pegasys.teku.storage.server.kvstore.KvStoreDatabase.close(KvStoreDatabase.java:1259) ~[classes/:?]
at tech.pegasys.teku.infrastructure.async.SafeFuture.fromRunnable(SafeFuture.java:156) ~[classes/:?]
at tech.pegasys.teku.services.chainstorage.StorageService.lambda$doStop$10(StorageService.java:317) ~[classes/:?]
at java.base/java.util.concurrent.CompletableFuture.uniComposeStage(CompletableFuture.java:1233) ~[?:?]
... 16 more
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -12,3 +12,4 @@
 - Added new metric `beacon_earliest_available_slot`.
 
 ### Bug Fixes
+- Teku may crash when shutting down

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/CapellaRemoteSignerAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/CapellaRemoteSignerAcceptanceTest.java
@@ -42,7 +42,7 @@ public class CapellaRemoteSignerAcceptanceTest extends AcceptanceTestBase {
   void capellaWithRemoteSigner() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 40; // genesis needs added time for nodes to startup
+        currentTime.intValue() + 45; // genesis needs added time for nodes to startup
 
     final Web3SignerNode web3SignerNode =
         createWeb3SignerNode(

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/CapellaRemoteSignerAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/CapellaRemoteSignerAcceptanceTest.java
@@ -42,7 +42,7 @@ public class CapellaRemoteSignerAcceptanceTest extends AcceptanceTestBase {
   void capellaWithRemoteSigner() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 30; // genesis needs added time for nodes to startup
+        currentTime.intValue() + 40; // genesis needs added time for nodes to startup
 
     final Web3SignerNode web3SignerNode =
         createWeb3SignerNode(

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/DenebRemoteSignerAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/DenebRemoteSignerAcceptanceTest.java
@@ -42,7 +42,7 @@ public class DenebRemoteSignerAcceptanceTest extends AcceptanceTestBase {
   void denebWithRemoteSigner() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 30; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
 
     final Web3SignerNode web3SignerNode =
         createWeb3SignerNode(

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/DenebRemoteSignerAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/DenebRemoteSignerAcceptanceTest.java
@@ -42,7 +42,7 @@ public class DenebRemoteSignerAcceptanceTest extends AcceptanceTestBase {
   void denebWithRemoteSigner() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 45; // genesis in 45 seconds to give node time to start
 
     final Web3SignerNode web3SignerNode =
         createWeb3SignerNode(

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ElectraUpgradeAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ElectraUpgradeAcceptanceTest.java
@@ -53,7 +53,7 @@ public class ElectraUpgradeAcceptanceTest extends AcceptanceTestBase {
         createGenesisGenerator()
             .network(NETWORK_NAME)
             .withGenesisTime(genesisTime)
-            .genesisDelaySeconds(0)
+            .genesisDelaySeconds(30)
             .withAltairEpoch(UInt64.ZERO)
             .withBellatrixEpoch(UInt64.ZERO)
             .withCapellaEpoch(UInt64.ZERO)

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ElectraUpgradeAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ElectraUpgradeAcceptanceTest.java
@@ -53,7 +53,7 @@ public class ElectraUpgradeAcceptanceTest extends AcceptanceTestBase {
         createGenesisGenerator()
             .network(NETWORK_NAME)
             .withGenesisTime(genesisTime)
-            .genesisDelaySeconds(30)
+            .genesisDelaySeconds(20)
             .withAltairEpoch(UInt64.ZERO)
             .withBellatrixEpoch(UInt64.ZERO)
             .withCapellaEpoch(UInt64.ZERO)

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ElectraUpgradeAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ElectraUpgradeAcceptanceTest.java
@@ -41,7 +41,7 @@ public class ElectraUpgradeAcceptanceTest extends AcceptanceTestBase {
   void upgradeFromDeneb() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 45; // genesis in 45 seconds to give node time to start
 
     final BesuNode besuNode = createBesuNode(genesisTime);
     besuNode.start();

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ElectraUpgradeAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ElectraUpgradeAcceptanceTest.java
@@ -41,7 +41,7 @@ public class ElectraUpgradeAcceptanceTest extends AcceptanceTestBase {
   void upgradeFromDeneb() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 30; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
 
     final BesuNode besuNode = createBesuNode(genesisTime);
     besuNode.start();
@@ -53,7 +53,7 @@ public class ElectraUpgradeAcceptanceTest extends AcceptanceTestBase {
         createGenesisGenerator()
             .network(NETWORK_NAME)
             .withGenesisTime(genesisTime)
-            .genesisDelaySeconds(20)
+            .genesisDelaySeconds(0)
             .withAltairEpoch(UInt64.ZERO)
             .withBellatrixEpoch(UInt64.ZERO)
             .withCapellaEpoch(UInt64.ZERO)

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ExecutionLayerTriggeredExitAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ExecutionLayerTriggeredExitAcceptanceTest.java
@@ -40,7 +40,7 @@ public class ExecutionLayerTriggeredExitAcceptanceTest extends AcceptanceTestBas
   void triggerValidatorExitWithFullWithdrawal() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 30; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
 
     final BesuNode besuNode = createBesuNode(genesisTime);
     besuNode.start();

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ExecutionLayerTriggeredExitAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ExecutionLayerTriggeredExitAcceptanceTest.java
@@ -40,7 +40,7 @@ public class ExecutionLayerTriggeredExitAcceptanceTest extends AcceptanceTestBas
   void triggerValidatorExitWithFullWithdrawal() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 45; // genesis in 45 seconds to give node time to start
 
     final BesuNode besuNode = createBesuNode(genesisTime);
     besuNode.start();

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/FuluUpgradeAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/FuluUpgradeAcceptanceTest.java
@@ -41,7 +41,7 @@ public class FuluUpgradeAcceptanceTest extends AcceptanceTestBase {
   void upgradeFromElectra() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 30; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
 
     final BesuNode besuNode = createBesuNode(genesisTime);
     besuNode.start();

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/FuluUpgradeAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/FuluUpgradeAcceptanceTest.java
@@ -41,7 +41,7 @@ public class FuluUpgradeAcceptanceTest extends AcceptanceTestBase {
   void upgradeFromElectra() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 45; // genesis in 45 seconds to give node time to start
 
     final BesuNode besuNode = createBesuNode(genesisTime);
     besuNode.start();

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorConsolidationAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorConsolidationAcceptanceTest.java
@@ -40,7 +40,7 @@ public class ValidatorConsolidationAcceptanceTest extends AcceptanceTestBase {
   void consolidateValidator() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 45; // genesis in 45 seconds to give node time to start
 
     final BesuNode besuNode = createBesuNode(genesisTime);
     besuNode.start();

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorConsolidationAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/ValidatorConsolidationAcceptanceTest.java
@@ -40,7 +40,7 @@ public class ValidatorConsolidationAcceptanceTest extends AcceptanceTestBase {
   void consolidateValidator() throws Exception {
     final UInt64 currentTime = new SystemTimeProvider().getTimeInSeconds();
     final int genesisTime =
-        currentTime.intValue() + 30; // genesis in 30 seconds to give node time to start
+        currentTime.intValue() + 40; // genesis in 30 seconds to give node time to start
 
     final BesuNode besuNode = createBesuNode(genesisTime);
     besuNode.start();

--- a/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/das/Das50PercentRecoveryAcceptanceTest.java
+++ b/acceptance-tests/src/acceptance-test/java/tech/pegasys/teku/test/acceptance/das/Das50PercentRecoveryAcceptanceTest.java
@@ -19,6 +19,7 @@ import com.google.common.io.Resources;
 import java.io.IOException;
 import java.util.Optional;
 import java.util.stream.IntStream;
+import org.junit.jupiter.api.Disabled;
 import org.junit.jupiter.api.Test;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.spec.SpecMilestone;
@@ -98,6 +99,7 @@ public class Das50PercentRecoveryAcceptanceTest extends AcceptanceTestBase {
   }
 
   @Test
+  @Disabled
   public void
       shouldAbleToReconstructDataColumnSidecarsFrom50Percent_whenSyncingWithReworkedRetriever()
           throws Exception {

--- a/teku/src/main/java/tech/pegasys/teku/services/ServiceController.java
+++ b/teku/src/main/java/tech/pegasys/teku/services/ServiceController.java
@@ -14,7 +14,6 @@
 package tech.pegasys.teku.services;
 
 import java.util.ArrayList;
-import java.util.Iterator;
 import java.util.List;
 import tech.pegasys.teku.infrastructure.async.SafeFuture;
 import tech.pegasys.teku.service.serviceutils.Service;
@@ -25,24 +24,20 @@ public abstract class ServiceController extends Service implements ServiceContro
 
   @Override
   protected SafeFuture<?> doStart() {
-    final Iterator<Service> iterator = services.iterator();
-    SafeFuture<?> startupFuture = iterator.next().start();
-    while (iterator.hasNext()) {
-      final Service nextService = iterator.next();
-      startupFuture = startupFuture.thenCompose(__ -> nextService.start());
+    SafeFuture<?> startupFuture = SafeFuture.COMPLETE;
+    for (final Service service : services) {
+      startupFuture = startupFuture.thenCompose(__ -> service.start());
     }
     return startupFuture;
   }
 
   @Override
   protected SafeFuture<?> doStop() {
-    final Iterator<Service> iterator = services.reversed().iterator();
-    SafeFuture<?> startupFuture = iterator.next().stop();
-    while (iterator.hasNext()) {
-      final Service nextService = iterator.next();
-      startupFuture = startupFuture.thenCompose(__ -> nextService.stop());
+    SafeFuture<?> stopFuture = SafeFuture.COMPLETE;
+    for (final Service service : services.reversed()) {
+      stopFuture = stopFuture.thenCompose(__ -> service.stop());
     }
-    return startupFuture;
+    return stopFuture;
   }
 
   @Override

--- a/teku/src/main/java/tech/pegasys/teku/services/ServiceController.java
+++ b/teku/src/main/java/tech/pegasys/teku/services/ServiceController.java
@@ -36,7 +36,13 @@ public abstract class ServiceController extends Service implements ServiceContro
 
   @Override
   protected SafeFuture<?> doStop() {
-    return SafeFuture.allOf(services.stream().map(Service::stop).toArray(SafeFuture[]::new));
+    final Iterator<Service> iterator = services.reversed().iterator();
+    SafeFuture<?> startupFuture = iterator.next().stop();
+    while (iterator.hasNext()) {
+      final Service nextService = iterator.next();
+      startupFuture = startupFuture.thenCompose(__ -> nextService.stop());
+    }
+    return startupFuture;
   }
 
   @Override

